File: backend/microservice1-1/abcd/src/main/java/com/example/demo/controller/BookingController.java
package com.example.demo.controller;

import java.util.Map;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;
import com.example.demo.model.Bookings;
import com.example.demo.service.BookingService;

@RestController
@RequestMapping("/booking")
public class BookingController {
    
    @Autowired
    BookingService bookingService;
 // Test endpoint - no token required
    @GetMapping("/test")
    public Object testEndpoint() {
        return Map.of("code", 200, "message", "Booking endpoint is working!");
    }
    @PostMapping("/book")
    public Object bookResource(@RequestBody Bookings booking, 
                               @RequestHeader(value = "Token", required = false) String token) {
        if (token == null || token.isEmpty()) {
            return Map.of("code", 401, "message", "Token is required");
        }
        return bookingService.bookResource(booking, token);
    }
    
    @GetMapping("/mybookings")
    public Object getMyBookings(@RequestHeader(value = "Token", required = false) String token) {
        System.out.println("Received token in controller: " + token);
        if (token == null || token.isEmpty()) {
            return Map.of("code", 401, "message", "Token is required. Please login first.");
        }
        return bookingService.getMyBookings(token);
    }
    
    @GetMapping("/all")
    public Object getAllBookings(@RequestHeader(value = "Token", required = false) String token) {
        if (token == null || token.isEmpty()) {
            return Map.of("code", 401, "message", "Token is required");
        }
        return bookingService.getAllBookings(token);
    }
    
    @GetMapping("/getbyresource/{resourceId}")
    public Object getBookingsByResource(@PathVariable("resourceId") int resourceId, 
                                        @RequestHeader(value = "Token", required = false) String token) {
        if (token == null || token.isEmpty()) {
            return Map.of("code", 401, "message", "Token is required");
        }
        return bookingService.getBookingsByResource(resourceId, token);
    }
    
    @PutMapping("/cancel/{bookingId}")
    public Object cancelBooking(@PathVariable("bookingId") int bookingId, 
                                @RequestHeader(value = "Token", required = false) String token) {
        if (token == null || token.isEmpty()) {
            return Map.of("code", 401, "message", "Token is required");
        }
        return bookingService.cancelBooking(bookingId, token);
    }
    
    @GetMapping("/get/{bookingId}")
    public Object getBookingById(@PathVariable("bookingId") int bookingId, 
                                 @RequestHeader(value = "Token", required = false) String token) {
        if (token == null || token.isEmpty()) {
            return Map.of("code", 401, "message", "Token is required");
        }
        return bookingService.getBookingById(bookingId, token);
    }
    
    @PutMapping("/update/{bookingId}")
    public Object updateBookingStatus(@PathVariable("bookingId") int bookingId, 
                                      @RequestParam("status") int status, 
                                      @RequestHeader(value = "Token", required = false) String token) {
        if (token == null || token.isEmpty()) {
            return Map.of("code", 401, "message", "Token is required");
        }
        return bookingService.updateBookingStatus(bookingId, status, token);
    }
}

File: backend/gateway/controllers/bookingController.py
from typing import Any

from fastapi import APIRouter, Header, Query, Response
import httpx

from controllers.proxy_utils import SPRING_BASE_URL, forward_response, to_spring_payload


router = APIRouter(prefix="/booking")
SPRING_BOOKING_URL = f"{SPRING_BASE_URL}/booking"


@router.get("/test")
async def test_booking(response: Response):
    async with httpx.AsyncClient() as client:
        spring_response = await client.get(f"{SPRING_BOOKING_URL}/test")
    return await forward_response(spring_response, response)


@router.post("/book")
async def book_resource(booking: dict[str, Any], response: Response, Token: str = Header(...)):
    async with httpx.AsyncClient() as client:
        spring_response = await client.post(
            f"{SPRING_BOOKING_URL}/book",
            json=to_spring_payload(booking),
            headers={"Token": Token},
        )
    return await forward_response(spring_response, response)


@router.get("/mybookings")
async def get_my_bookings(response: Response, Token: str = Header(...)):
    async with httpx.AsyncClient() as client:
        spring_response = await client.get(
            f"{SPRING_BOOKING_URL}/mybookings",
            headers={"Token": Token},
        )
    return await forward_response(spring_response, response)


@router.get("/all")
async def get_all_bookings(response: Response, Token: str = Header(...)):
    async with httpx.AsyncClient() as client:
        spring_response = await client.get(
            f"{SPRING_BOOKING_URL}/all",
            headers={"Token": Token},
        )
    return await forward_response(spring_response, response)


@router.get("/getbyresource/{resource_id}")
async def get_bookings_by_resource(
    resource_id: int,
    response: Response,
    Token: str = Header(...),
):
    async with httpx.AsyncClient() as client:
        spring_response = await client.get(
            f"{SPRING_BOOKING_URL}/getbyresource/{resource_id}",
            headers={"Token": Token},
        )
    return await forward_response(spring_response, response)


@router.put("/cancel/{booking_id}")
async def cancel_booking(booking_id: int, response: Response, Token: str = Header(...)):
    async with httpx.AsyncClient() as client:
        spring_response = await client.put(
            f"{SPRING_BOOKING_URL}/cancel/{booking_id}",
            headers={"Token": Token},
        )
    return await forward_response(spring_response, response)


@router.get("/get/{booking_id}")
async def get_booking_by_id(booking_id: int, response: Response, Token: str = Header(...)):
    async with httpx.AsyncClient() as client:
        spring_response = await client.get(
            f"{SPRING_BOOKING_URL}/get/{booking_id}",
            headers={"Token": Token},
        )
    return await forward_response(spring_response, response)


@router.put("/update/{booking_id}")
async def update_booking_status(
    booking_id: int,
    response: Response,
    status: int = Query(...),
    Token: str = Header(...),
):
    async with httpx.AsyncClient() as client:
        spring_response = await client.put(
            f"{SPRING_BOOKING_URL}/update/{booking_id}",
            params={"status": status},
            headers={"Token": Token},
        )
    return await forward_response(spring_response, response)
